add allowed users to tsh db ls json and yaml output

GavinFrazar opened the pull request:
Closes #16430

## What
When a user runs `tsh db ls --format=json` or `--format=yaml`, the output should include the user's allowed/denied db_users, just like it does for `--format=text`. This PR adds db_users info to json/yaml output.

## How
Combine each database with the user's allowed/denied db_users in a new struct before marshaling, so that each database is printed under a new `database:` field alongside a new `db_users:` field.

## Tests?
Tests cover various RoleSet combinations, and each of text/json/yaml format flag. Tests also cover a `nil` RoleSet scenario, although this should never happen in practice, but I handled that scenario by falling back to the old format with just the databases.

## Discussion
Should we fall back to the old format for databases if `nil` RoleSet is given? Or should we just make `db_users` nil and still print each database under the new `database:` field for consistent format? Alternatively to both, we could just return an error, as this is currently unexpected to ever happen. I have no preference really.

## Output diff
(ignoring the extra indentation whitespace for the database):
```diff
+- database:
    kind: db
    metadata:
      description: Example dyn database resource
      labels:
        config: dynamic
        region: local
        teleport.dev/origin: dynamic
      name: home-db-psql
    spec:
      ad:
        domain: ""
        keytab_file: ""
        spn: ""
      aws:
        elasticache: {}
        memorydb: {}
        rds:
          iam_auth: false
        rdsproxy: {}
        redshift: {}
        secret_store: {}
      azure:
        redis: {}
      gcp: {}
      mysql: {}
      protocol: postgres
      tls:
        mode: 0
      uri: postgres.mars.local.gd:5432
    status:
      aws:
        elasticache: {}
        memorydb: {}
        rds:
          iam_auth: false
        rdsproxy: {}
        redshift: {}
        secret_store: {}
      azure:
        redis: {}
      mysql: {}
    version: v3
+  db_users:
+    allowed:
+    - '*'
+    - devUser
+    - prodUser
+    denied:
+    - ROOT
+    - baduser
```

GavinFrazar committed:
Output allowed/denied users in tsh db ls --format [json | yaml]

GavinFrazar committed:
Add tests for new output

* tested with nil role set
* tested with multiple role combos
* tested yaml and json output

espadolini commented:
Do we need the object to match exactly the structure of a `types.DatabaseV3`? I'd rather not change the format of what we're currently outputting and perhaps just pretend that there's a `users` field in it.

GavinFrazar commented:
> Do we need the object to match exactly the structure of a `types.DatabaseV3`? I'd rather not change the format of what we're currently outputting and perhaps just pretend that there's a `users` field in it.

Well, I could inline the database and then add a `db_users` field in that, but that inlining doesn't work with an embedded interface so I would need to use a map and commit some reflect crimes... or some inefficient json marshal/unmarshal like this:

```go
		blob, err := json.Marshal(db)
		if err != nil {
			return nil, trace.Wrap(err)
		}
		var dbOutput map[string]any
		json.Unmarshal(blob, &dbOutput)
		dbOutput["db_users"] = &DbUsers{
			Allowed: allowed,
			Denied:  denied,
		}
		result = append(result, dbOutput)
```

Prints like this:
```
- db_users:
    allowed:
    - '*'
    denied: []
  kind: db
  metadata:
    description: Azure MySQL server in eastus
    labels:
      engine: Microsoft.DBforMySQL/servers
      engine-version: "5.7"
      foo: bar
      region: eastus
      resource-group: database-access
      subscription-id: 060a97ea-3a57-4218-9be5-dba3f19ff2b5
      teleport.dev: baz
      teleport.dev/origin: cloud
    name: teleport-mysql
...
```

espadolini commented:
We do the marshal-unmarshal thing so often that we have a utility function for it already (`api/utils.ObjectToStruct`). 😬 

It's also not very clear to me why `tsh db ls --format=json` should mimic the output that's already available from `tctl get db` rather than have a more friendly representation that can include extra fields more easily - do we really want our users to have to remember the difference between `metadata` and `spec`, or to see a whole bunch of empty maps like in your output?

GavinFrazar commented:
> We do the marshal-unmarshal thing so often that we have a utility function for it already (`api/utils.ObjectToStruct`). 😬
> 
> It's also not very clear to me why `tsh db ls --format=json` should mimic the output that's already available from `tctl get db` rather than have a more friendly representation that can include extra fields more easily - do we really want our users to have to remember the difference between `metadata` and `spec`, or to see a whole bunch of empty maps like in your output?

It's not clear to me either, but that is the way we've had it... I don't see why we shouldn't match the --format=text information, which doesn't include the full spec and everything. So I basically agree with you, but the other info is outside the scope of the issue addressed by this change. This isn't an urgent issue, so we can wait to see if additional scope should be added before approving this PR.

@smallinsky @r0mant what are your thoughts?

smallinsky commented:
@GavinFrazar 
> Should we just make the yaml/json versions of this match those headings?

I don't this that this is necessary. I think that for now we want to make sure that the json/yaml version after adding `db_users` is backward compatible and won't break if `tsh db ls --yaml/json` format is used by third party tool/script so i think that we can create a separate object and inline db object with additional db_users information for now.

GavinFrazar committed:
Clean up conditional logic in serializeDatabases

GavinFrazar committed:
Omit nil allowed/denied fields from database serialized info

GavinFrazar committed:
Refactor db users construction

GavinFrazar committed:
Fix tests

GavinFrazar commented:
@smallinsky @espadolini I refactored a bit and added a test case with the expected string output to verify it prints the db_users field inline, as if it were a field of the database:
```
- db_users:
    allowed:
    - '*'
    denied: []
  kind: db
  metadata:
    description: Azure MySQL server in eastus
    labels:
      engine: Microsoft.DBforMySQL/servers
      engine-version: "5.7"
      foo: bar
      region: eastus
      resource-group: database-access
      subscription-id: 060a97ea-3a57-4218-9be5-dba3f19ff2b5
      teleport.dev: baz
      teleport.dev/origin: cloud
    name: teleport-mysql
...
```

Please take another look.

smallinsky commented on tool/tsh/tsh.go:
```diff
@@ -2249,20 +2249,99 @@ func showDatabases(w io.Writer, clusterFlag string, databases []types.Database,
+
+func serializeDatabasesWithUsers(dbsWithUsers []*databaseWithUsers, format string) (string, error) {
+	// initialize to an empty slice so we don't print `null` in JSON/YAML if the cluster has no databases.
+	if dbsWithUsers == nil {
+		dbsWithUsers = []*databaseWithUsers{}
+	}
```
This function only differes from `serializeDatabases` end flow by this call. 
Instead of returning raw string we can make a helper function something like like 
```
func getDatabasePrintObject(databases, ...) any, error {
       if roleSet != nil {
		dbsWithUsers, err := getDatabasesWithUsers(databases, roleSet)
		if err != nil {
			return "", trace.Wrap(err)
		}
		return dbsWithUsers
	}
       return databases
}
```
and call `getDatabasePrintObject` function without adding duplication

GavinFrazar committed:
Refactor to deduplicate code

GavinFrazar committed:
Refactor TestListDatabasesWithUsers into a single table test

GavinFrazar committed:
Add more database serialization tests

GavinFrazar commented on tool/tsh/tsh_test.go:
```diff
@@ -3248,60 +3281,136 @@ func Test_getUsersForDb(t *testing.T) {
+			testName := fmt.Sprintf("%s %s", f, tt.name)
+			t.Run(testName, func(t *testing.T) {
+				t.Parallel()
+				out, err := serializeDatabases([]types.Database{tt.database}, f, tt.roles)
+				require.NoError(t, err)
+
+				var want []byte
+				switch f {
+				case teleport.JSON:
+					want, err = utils.FastMarshalIndent(tt.wantFormatted, "", "  ")
+				case teleport.YAML:
+					want, err = yaml.Marshal(tt.wantFormatted)
+				}
+				require.NoError(t, err)
+				require.Empty(t, cmp.Diff(string(want), string(out)))
```
I refactored this test to be one test table loop

GavinFrazar committed:
Change db_users to users for better sorted order db ls

GavinFrazar committed:
Rename DBUsers -> Users

GavinFrazar committed:
Merge branch 'master' into gavinfrazar/add-allowed-users-to-tsh-db-ls-json-yaml-output

GavinFrazar commented:
Actually, since this is a feature request I don't think we needed to backport it. I already backported to v11, but I'm just going to skip v9 and v10 backports.
